department-of-veterans-affairs/caseflow
# CASEFLOW-1617 Remove Room Limits on Building Dockets And Old Build Schedule Algorithm

Resolves [CASEFLOW-1617](https://vajira.max.gov/browse/CASEFLOW-1617)

### Description
This PR removes the old algorithm for bulk building dockets. It also removes all `room` based limitations to that bulk build process.

Before this PR we have two different algorithms in use when bulk building dockets (hearing_days) for ROs.
1. Video days (which have a `room`) are allocated to ros by an older algorithm with the entry point `allocate_hearing_days_to_ros`.
2. Virtual days (which have no `room`) are allocated to ros by a new algorithm with the entry point `allocate_no_room_hearing_days_to_ros`.

After this PR we have one algorithm with the entry point `allocate_hearing_days_to_ros`. This is the same algorithm as `allocate_no_room_hearing_days_to_ros` but the name no longer made sense.

### Acceptance Criteria
- [ ] Any RO can have as many dockets built for it as specified in the uploaded spreadsheet. No more hard-coded DOUBLE_ROOM_ROS and TRIPLE_ROOM_ROS)
- [ ] Rooms are no longer assigned to Video or Central hearing days
    - I found this easiest to see in the UI.
    - If you follow the `binding.pry` instructions below you should also be able to do something like this.
```ruby
displayed_hearing_days.select { |hd| (hd.request_type == "V" || hd.request_type == "R") }.pluck(:room).uniq
```
- [ ] Distribute the total number of Video and Virtual days for each RO evenly across the date range
    - The test for this is written to pass when this is +/-5 days per date. That's the same as the other evenness requirement. 
    - Here's the test that checks this: `it "allocates each type of days evenly across available dates for each ro"`
- [ ] Distribute the total number of Video and Virtual days equally across the date range (+/-5 days per date)
    - Here's the test that checks this: `it "allocates all days evenly across available dates for each ro" do`
    - I'm not 100% sure this test looks at the correct thing? I believe it does.

### Additional Work
Because we're removing the concept of a room from the video days that are built by this algorithm, it would make the code a lot clearer to rename two columns of the allocations table:
```ruby
:allocated_days # This should be renamed 'allocated_video_days'
:allocated_days_without_room # This should be renamed 'allocated_virtual_days'
# Usage of these can be seen at the entry points here: ro_schedule_period.rb::generate_ro_hearing_schedule
```
That work requires a migration and this PR is already fairly large, so this will need to be done in another PR.

### Testing Plan
Copied, then added to the plan on [PR 16193](https://github.com/department-of-veterans-affairs/caseflow/pull/16193).
- To replicate PROD and exclude Travel board days, comment out the call to method `filter_travel_board_hearing_days`
- Shadow user `BVASYELLOW` and navigate to the build schedule page: http://localhost:3000/hearings/schedule/build
- Click the Upload Files button
- Set the start date to 04/01/2021 and the end date to 06/30/2021
- Upload this file: `spec/support/q3fy21RealExampleRoSpreadsheet.xlsx`
- Ensure that the algorithm runs successfully, then try the `Console exploration` or `Pivot Table` sections below. Those sections allow testing the same things (evenness of distribution of hearing_days) in different ways.

####  Console Exploration Notes
- This PR adds some fairly complex tests so that we can test business logic, see `generate_hearing_days_schedule_spec.rb`
- Those tests make it easier to see what's happening when that file is uploaded because it's used as a test case.
- Add a `binding.pry` to the block under `let :day_counts_and_summary_per_ro do` in `generate_hearing_days_schedule_spec.rb` and you'll be able to examine both the `:ro_schedule_period` that results from that spreadsheet and the various summary data the test uses.
- Then run `make one-test spec/services/hearing_schedule/generate_hearing_days_schedule_spec.rb`
- You can then examine the data by doing things like this `day_counts_and_summary_per_ro["RO17"]["summary_statistics"]`. The summary statistics are designed to give you the numbers that we would be scanning the pivot table for.

#### (Deprecated) Create the Distribution Pivot Table
- Once the algorithm runs successfully and you are presented with the results in a table, copy everything including the headers and paste into a new Excel workbook
- In Excel, with the whole data range selected (headers included) go to the "Insert" tab and click on the "Pivot Table". Add the table to a new worksheet to start building the pivot table
- Drag the Regional Office to the columns, the Date to the rows and the values (Count) and then drag the rooms to the filter
- Apply the filter removing all rooms except the blank ones
- View the results, make sure to expand the months to see each days distribution
